### PR TITLE
fix: catch listen error

### PR DESCRIPTION
**Pull request by mpetrunic**
This should fix it but for whatever reason it's not working. Please help.

Related to: https://github.com/libp2p/js-libp2p-websockets/issues/184#event-7914439705

### CODE REVIEW

**Comment by mpetrunic** [comment]
cc @achingbrain

**Comment by achingbrain** [comment]
The reason this is failing is because the websocket server [forwards any error events](https://github.com/websockets/ws/blob/master/lib/websocket-server.js#L112) emitted by the underlying http(s) server when it's passed as an option.

If you stick a `wsServer.on('error', () => {})` after `wsServer` [is created](https://github.com/alanshaw/it-ws/blob/master/src/server.ts#L72-L76) the test passes (almost, the message assertion needs updating to something like `await expect(server2.listen(55215)).to.eventually.be.rejected.with.property('message').that.include('EADDRINUSE')`).

We probably want to be a bit smarter about error handling than that though.

We could update the internal `Server` class to create the`WSServer` instance in the `listen` method, call `.listen` on the underlying http(s) server but resolve the promise on the `WSServer` events?

**Comment by mpetrunic** [comment]
@achingbrain It works now. Hope I didn't do anything stupid 😅

File: src/server.ts [hunk 1]
```diff
@@ -22,15 +22,24 @@ export interface WebSocketServer extends EventEmitter {
...
+    this.wsServer.on('connection', this.onWsServerConnection.bind(this))
   }
 
   async listen (addrInfo: { port: number } | number) {
-    return await new Promise<WebSocketServer>(resolve => {
+    return await new Promise<WebSocketServer>((resolve, reject) => {
+      this.server.once('error', (e) => reject(e))
+      this.wsServer.once('error', (e) => reject(e))
       this.once('listening', () => resolve(this))
       this.server.listen(typeof addrInfo === 'number' ? addrInfo : addrInfo.port)
```
Comment (achingbrain) [inline src/server.ts:44]: The WSServer sets up an error listener and also forwards on the listen events from the underlying http(s) server so we can just listen for events on that to resolve the promise.

```suggestion
    return await new Promise<WebSocketServer>((resolve, reject) => {
      this.wsServer.once('error', (e) => reject(e))
      this.wsServer.once('listening', () => resolve(this))
      this.server.listen(typeof addrInfo === 'number' ? addrInfo : addrInfo.port)
```

**Review by achingbrain — APPROVED** [review]

**Pull request merged**